## Analitica Computacional para la toma de Decisiones
## Proyecto 2 - Productos bancarios
Carlos Gomez - 202111593
<br>
Jeronimo Vargas - 202113305
<br>
Juan Esteban Lopez - 202021417
<br>

In [16]:
# Importar librerias
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns
import tensorflow as tf

from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.datasets import load_iris

from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score

# Importar el dataset
dataset = pd.read_csv('bank-full-clean.csv')

### 1er Modelo de Redes neuronales

Separacion datos de entrenamiento y prueba

In [17]:
train = dataset.sample(frac=0.8, random_state=100)
train.head()

# Preprocesamiento de los datos
X = dataset.drop(['y'], axis=1)
y = dataset['y']

# Dividir el dataset en conjunto de entrenamiento y conjunto de prueba
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

# Escalado de características
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

X_train = np.array(X_train)
X_test = np.array(X_test)
y_train = np.array(y_train)
y_test = np.array(y_test)

# Shape
print(X_train.shape)
print(X_test.shape)
print(y_train.shape)
print(y_test.shape)


(36168, 23)
(9043, 23)
(36168,)
(9043,)


### Modelo 1: Red más profunda con más capas ocultas y el optimizador SGD

In [18]:
from tensorflow.keras import layers, models

# Definimos el modelo
model1 = models.Sequential()

# Capas ocultas adicionales
model1.add(layers.Dense(128, activation='relu', input_shape=(X_train.shape[1],)))
model1.add(layers.Dense(64, activation='relu'))
model1.add(layers.Dense(32, activation='relu'))
model1.add(layers.Dense(16, activation='relu'))

# Capa de salida
model1.add(layers.Dense(1, activation='sigmoid'))

# Compilamos el modelo con el optimizador SGD
model1.compile(optimizer='sgd', loss='binary_crossentropy', metrics=['accuracy'])

# Entrenamos el modelo
history1 = model1.fit(X_train, y_train, epochs=100, batch_size=32, validation_split=0.2)

# Resumen del modelo
model1.summary()

c:\Users\Jeronimo Vargas\AppData\Local\Programs\Python\Python311\Lib\site-packages\keras\src\layers\core\dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/100
905/905 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8861 - loss: 0.3700 - val_accuracy: 0.8913 - val_loss: 0.2499
Epoch 2/100
905/905 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8944 - loss: 0.2511 - val_accuracy: 0.9052 - val_loss: 0.2300
Epoch 3/100
905/905 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8992 - loss: 0.2415 - val_accuracy: 0.9066 - val_loss: 0.2251
Epoch 4/100
905/905 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9009 - loss: 0.2348 - val_accuracy: 0.9078 - val_loss: 0.2231
Epoch 5/100
905/905 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8998 - loss: 0.2327 - val_accuracy: 0.9050 - val_loss: 0.2233
Epoch 6/100
905/905 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9041 - loss: 0.2271 - val_accuracy: 0.9049 - val_loss: 0.2223
Epoch 7/100
905/905 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9030 - loss: 0.2301 - val_accuracy: 0.9066 - val_loss: 0.2202
Epoch 8/100
905/905 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9071 - loss: 0.2243 - val_accu

Model: "sequential_4"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_14 (Dense)                │ (None, 128)            │         3,072 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_15 (Dense)                │ (None, 64)             │         8,256 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_16 (Dense)                │ (None, 32)             │         2,080 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_17 (Dense)                │ (None, 16)             │           528 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_18 (Dense)                │ (None, 1)              │            17 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 13,955 (54.52 KB)

 Trainable params: 13,953 (54.50 KB)

 Non-trainable params: 0 (0.00 B)

 Optimizer params: 2 (12.00 B)

### Modelo 2: Red más ancha con Adam y función de activación tanh

In [19]:
from tensorflow.keras import layers, models

# Definimos el modelo
model2 = models.Sequential()

# Capas ocultas con activación tanh
model2.add(layers.Dense(256, activation='tanh', input_shape=(X_train.shape[1],)))
model2.add(layers.Dense(128, activation='tanh'))
model2.add(layers.Dense(64, activation='tanh'))

# Capa de salida
model2.add(layers.Dense(1, activation='sigmoid'))

# Compilamos el modelo con el optimizador Adam
model2.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

# Entrenamos el modelo
history2 = model2.fit(X_train, y_train, epochs=100, batch_size=32, validation_split=0.2)

# Resumen del modelo
model2.summary()


Epoch 1/100
905/905 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.8759 - loss: 0.2920 - val_accuracy: 0.9001 - val_loss: 0.2342
Epoch 2/100
905/905 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8989 - loss: 0.2338 - val_accuracy: 0.9056 - val_loss: 0.2170
Epoch 3/100
905/905 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8989 - loss: 0.2282 - val_accuracy: 0.9053 - val_loss: 0.2148
Epoch 4/100
905/905 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.9055 - loss: 0.2176 - val_accuracy: 0.9050 - val_loss: 0.2150
Epoch 5/100
905/905 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.9089 - loss: 0.2089 - val_accuracy: 0.9014 - val_loss: 0.2158
Epoch 6/100
905/905 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.9078 - loss: 0.2072 - val_accuracy: 0.8994 - val_loss: 0.2140
Epoch 7/100
905/905 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.9099 - loss: 0.2023 - val_accuracy: 0.9050 - val_loss: 0.2128
Epoch 8/100
905/905 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.9151 - loss: 0.1931 - val_accu

Model: "sequential_5"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_19 (Dense)                │ (None, 256)            │         6,144 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_20 (Dense)                │ (None, 128)            │        32,896 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_21 (Dense)                │ (None, 64)             │         8,256 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_22 (Dense)                │ (None, 1)              │            65 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 142,085 (555.02 KB)

 Trainable params: 47,361 (185.00 KB)

 Non-trainable params: 0 (0.00 B)

 Optimizer params: 94,724 (370.02 KB)

### Modelo 3: Red más simple con el optimizador RMSprop

In [20]:
from tensorflow.keras import layers, models

# Definimos el modelo
model3 = models.Sequential()

# Capas ocultas con activación relu
model3.add(layers.Dense(32, activation='relu', input_shape=(X_train.shape[1],)))
model3.add(layers.Dense(16, activation='relu'))

# Capa de salida
model3.add(layers.Dense(1, activation='sigmoid'))

# Compilamos el modelo con el optimizador RMSprop
model3.compile(optimizer='rmsprop', loss='binary_crossentropy', metrics=['accuracy'])

# Entrenamos el modelo
history3 = model3.fit(X_train, y_train, epochs=100, batch_size=32, validation_split=0.2)

# Resumen del modelo
model3.summary()


Epoch 1/100
905/905 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8453 - loss: 0.3530 - val_accuracy: 0.9056 - val_loss: 0.2279
Epoch 2/100
905/905 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.9017 - loss: 0.2283 - val_accuracy: 0.9050 - val_loss: 0.2252
Epoch 3/100
905/905 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.9030 - loss: 0.2310 - val_accuracy: 0.9035 - val_loss: 0.2221
Epoch 4/100
905/905 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9040 - loss: 0.2273 - val_accuracy: 0.9050 - val_loss: 0.2225
Epoch 5/100
905/905 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.9046 - loss: 0.2268 - val_accuracy: 0.9043 - val_loss: 0.2241
Epoch 6/100
905/905 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.9025 - loss: 0.2293 - val_accuracy: 0.9049 - val_loss: 0.2209
Epoch 7/100
905/905 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.9055 - loss: 0.2220 - val_accuracy: 0.9046 - val_loss: 0.2209
Epoch 8/100
905/905 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.9061 - loss: 0.2233 - val_accu

Model: "sequential_6"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_23 (Dense)                │ (None, 32)             │           768 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_24 (Dense)                │ (None, 16)             │           528 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_25 (Dense)                │ (None, 1)              │            17 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 2,628 (10.27 KB)

 Trainable params: 1,313 (5.13 KB)

 Non-trainable params: 0 (0.00 B)

 Optimizer params: 1,315 (5.14 KB)

In [21]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Evaluar el Modelo 1
y_pred_1 = (model1.predict(X_test) > 0.5).astype("int32")
accuracy_1 = accuracy_score(y_test, y_pred_1)
precision_1 = precision_score(y_test, y_pred_1)
recall_1 = recall_score(y_test, y_pred_1)
f1_1 = f1_score(y_test, y_pred_1)

# Evaluar el Modelo 2
y_pred_2 = (model2.predict(X_test) > 0.5).astype("int32")
accuracy_2 = accuracy_score(y_test, y_pred_2)
precision_2 = precision_score(y_test, y_pred_2)
recall_2 = recall_score(y_test, y_pred_2)
f1_2 = f1_score(y_test, y_pred_2)

# Evaluar el Modelo 3
y_pred_3 = (model3.predict(X_test) > 0.5).astype("int32")
accuracy_3 = accuracy_score(y_test, y_pred_3)
precision_3 = precision_score(y_test, y_pred_3)
recall_3 = recall_score(y_test, y_pred_3)
f1_3 = f1_score(y_test, y_pred_3)

# Imprimir los resultados
print("Modelo 1")
print(f"Accuracy: {accuracy_1}, Precision: {precision_1}, Recall: {recall_1}, F1 Score: {f1_1}")
print("Modelo 2")
print(f"Accuracy: {accuracy_2}, Precision: {precision_2}, Recall: {recall_2}, F1 Score: {f1_2}")
print("Modelo 3")
print(f"Accuracy: {accuracy_3}, Precision: {precision_3}, Recall: {recall_3}, F1 Score: {f1_3}")

# Crear tabla con los datos anteriores
data = {
    'Modelo': ['Modelo 1', 'Modelo 2', 'Modelo 3'],
    'Accuracy': [accuracy_1, accuracy_2, accuracy_3],
    'Precision': [precision_1, precision_2, precision_3],
    'Recall': [recall_1, recall_2, recall_3],
    'F1 Score': [f1_1, f1_2, f1_3]
}

df = pd.DataFrame(data)
df

283/283 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step
283/283 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step
283/283 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step
Modelo 1
Accuracy: 0.8915183014486343, Precision: 0.5753676470588235, Recall: 0.2944496707431797, F1 Score: 0.3895457373988799
Modelo 2
Accuracy: 0.8849939179475838, Precision: 0.5120925341745531, Recall: 0.458137347130762, F1 Score: 0.48361469712015887
Modelo 3
Accuracy: 0.894061705186332, Precision: 0.5816485225505443, Recall: 0.35183443085606775, F1 Score: 0.4384525205158265


Modelo  Accuracy  Precision    Recall  F1 Score
0  Modelo 1  0.891518   0.575368  0.294450  0.389546
1  Modelo 2  0.884994   0.512093  0.458137  0.483615
2  Modelo 3  0.894062   0.581649  0.351834  0.438453

Modelo 1 tiene una buena precisión (0.575), pero un recall bajo (0.294), lo cual indica que predice correctamente a algunos clientes que se suscribirán, pero pierde a muchos otros. Puede no ser la mejor opción si el objetivo es alcanzar a la mayor cantidad posible de clientes interesados.

Modelo 2 tiene el mejor balance entre precisión y recall, con un F1 Score de 0.483. Este modelo captura correctamente a más clientes que realmente se suscribirán (recall de 0.458), aunque también incluye algunos falsos positivos. Puede ser el mejor modelo si el área comercial quiere un equilibrio entre minimizar falsos positivos y no perder clientes potencialmente interesados.

Modelo 3 tiene la precisión más alta (0.581), lo que significa que sus predicciones positivas son más confiables, pero su recall es menor en comparación con el Modelo 2. Si la prioridad es evitar falsos positivos (es decir, predecir que un cliente se suscribirá solo cuando sea realmente probable), este modelo es adecuado. Sin embargo, sacrificará algo de cobertura, ya que no identificará a todos los clientes interesados.

Dado el contexto y la prioridad en la conversión de clientes, el Modelo 2 parece ser la mejor opción, ya que ofrece un equilibrio sólido entre precisión y recall, lo cual es útil para maximizar la conversión sin aumentar demasiado el número de falsos positivos.

In [26]:
# Guardar el segundo modelo para poder predecir en el futuro
model2.save('model2.h5')

# Guardar el Xtrain y el ytrain para poder predecir en el futuro
np.save('X_train.npy', X_train)

In [ ]:
# Ejemplo de datos de entrada (asegúrate de tener la misma cantidad de columnas que en el entrenamiento)
# Este ejemplo asume que tienes los datos de un solo cliente en un array
nuevo_cliente = np.array([[40, 0, 1500, 1, 0, 15, 5, 300, 2, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0]])

# Escalar los datos usando el mismo scaler que se usó en el entrenamiento
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)  # Solo si no tienes el scaler guardado, entrena con X_train
nuevo_cliente_scaled = scaler.transform(nuevo_cliente)

# Importar el modelo
model_imp = tf.keras.models.load_model('model2.h5')

# Realizar la predicción
prediccion = model_imp.predict(nuevo_cliente_scaled)

# Clasificar la predicción (0 o 1)
resultado = (prediccion > 0.5).astype("int32")

# Imprimir el resultado
print(f"Probabilidad de conversión: {prediccion[0][0]:.2f}")
print(f"Conversión predicha (1 = sí, 0 = no): {resultado[0][0]}")

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 78ms/step
Probabilidad de conversión: 1.00
Conversión predicha (1 = sí, 0 = no): 1
